README.md
Wrapper for web_socket_channel package lightweight and isomorphic lib with socket.io-like event handling, Future-based requests. This wrapper implements the protocol for [this ws node wrapper](https://github.com/bminer/ws-wrapper).

**WARNING: server -> client requests are not sopported. If someone needs them open a PR**

## Usage

A simple usage example:

**Client side:**
```dart
import 'dart:async';
import 'dart:io';
import 'package:websocket_channel_wrapper/websocket_channel_wrapper.dart';

const PORT = 42069;

main() {
  WebSocketChannelWrapper socket = WebSocketChannelWrapper('ws://192.168.0.18:$PORT',
                                              headers: {'id': '1234567890qwertyuiop'});

  print('Connecting...');

  var onReady = socket.onConnect.first;

  onReady.timeout(Duration(seconds: 5), onTimeout: () => print('Server could not be reach, trying to re-connect'));

  socket.onConnect.skip(1).listen((_) => print('Successful re-connection to server'));

  socket.onDone.listen((_) => print(socket.autoReconnect ? 'Connection lost, trying to re-connect!' : 'Connection ended!'));

  onReady.then((_) async { // [onConnect] It's called every time the WebSocket reconnect
    print('Connected!');

    socket.emit('msg', ['DART_WEBSOCKET_WRAPPER', 'Hello, World!']);

    socket.on('serverTime').listen((time) => print('Time: $time'));  // Time: 2019-07-29T14:00:32.635Z

    socket.request('userCount').then((n) => print('# users: $n'));   // # users: 1

    socket.request('checkError').catchError((e) => print(e));        // Yep, errors work

    var filename = '/path/to/file/image.jpg';
    sendFile(filename, socket);   // Send binary file over socket

    Timer(Duration(seconds: 10), () {    // Close the socket after 1 minute
      var reason = 'break time';
      print('Close reason: $reason');
      socket.close(closeReason: reason);
    });
  });
}

sendFile(String filename, socket) async {
  var image = File(filename);
  var contents = await image.readAsBytes();

  final stopwatch = Stopwatch()..start();   // Start timer to measure speed
  
  socket.request('transferFile', [filename, contents]).then((_) { // Send first file name then the bytes

    final speed = contents.length / stopwatch.elapsedMicroseconds;
    print('File sended in ${stopwatch.elapsed} at $speed MB/s');  // File sended in 0:00:00.058191 at 1.63 MB/s

  }).catchError((e) => print(e));
}


```

**Server side:**
```js
const WebSocketServer = require("ws").Server
  , WebSocketWrapper = require("ws-wrapper")
const fs = require('fs')

const PORT = 42069

var wss = new WebSocketServer({port: PORT});
var sockets = new Set();

wss.on("connection", (sckt, req) => {
  const socket = new WebSocketWrapper(sckt)
  sockets.add(socket)

  console.log('New socket connected')
  socket.emit('serverTime', new Date())  // Send data to client on connect
  
  socket.on("msg", function(from, msg) {
      console.log(`Received message from ${from}: ${msg}`) // Received message from DART_WEBSOCKET_WRAPPER: Hello, World!
  });

  socket.on("userCount", () => {  // Request
  return sockets.size;
  });
  
  socket.on("checkError", () => { // Request rejecting the Promise
      throw 'Yep, errors work'
  })

  socket.on("transferFile", (filename, buffer) => {
    return new Promise((resolve, reject) => {
      fs.writeFile(filename, Buffer.from(buffer), function(err) { // Save binary data recieved to file
        if(err)
          reject(err)
        else{
          resolve(true)
          console.log('File received and saved!')
        }
      })
    })
  })

  socket.on("disconnect", (event) => {
      console.log(`REASON: ${event.reason} CODE: ${event.code}`) // REASON: break time    CODE: 1007
      sockets.delete(socket);
  });
})

console.log('Listening on port: ' + PORT)
```


CHANGELOG.md
## 1.1.4

- Official support for binary files
- Expose socket `autoReconnect` attribute
- Default `closeCode` value is 1000 = `NORMAL_CLOSURE`

## 1.1.3

- Fix bug on events stream e.g. `onConnect` & `onDone`
- Improving example code

## 1.1.2

- Pedantic dependencies fix for flutter

## 1.1.1

- Bugfix on `close` method

## 1.1.0

- Add auto-connect feature
- Add `onDone` stream-events
- Add `onConnect` stream-events, this includes reconnect events

## 1.0.1

- Fix health package issues

## 1.0.0

- Initial version, created by cTatu


example/node_ws_server_example/server.js
const WebSocketServer = require("ws").Server
  , WebSocketWrapper = require("ws-wrapper")
const fs = require('fs')

const PORT = 42069

var wss = new WebSocketServer({port: PORT});
var sockets = new Set();

wss.on("connection", (sckt, req) => {
  const socket = new WebSocketWrapper(sckt)
  sockets.add(socket)

  console.log('New socket connected')
  socket.emit('serverTime', new Date())  // Send data to client on connect
  
  socket.on("msg", function(from, msg) {
      console.log(`Received message from ${from}: ${msg}`) // Received message from DART_WEBSOCKET_WRAPPER: Hello, World!
  });

  socket.on("userCount", () => {  // Request
  return sockets.size;
  });
  
  socket.on("checkError", () => { // Request rejecting the Promise
      throw 'Yep, errors work'
  })

  socket.on("transferFile", (filename, buffer) => {
    return new Promise((resolve, reject) => {
      fs.writeFile(filename, Buffer.from(buffer), function(err) { // Save binary data recieved to file
        if(err)
          reject(err)
        else{
          resolve(true)
          console.log('File received and saved!')
        }
      })
    })
  })

  socket.on("disconnect", (event) => {
      console.log(`REASON: ${event.reason} CODE: ${event.code}`) // REASON: break time    CODE: 1007
      sockets.delete(socket);
  });
})

console.log('Listening on port: ' + PORT)

lib/src/websocket_channel_wrapper_base.dart
import 'dart:async';
import 'dart:collection';
import 'dart:io';
import 'dart:convert';
import 'dart:math';
import 'package:pedantic/pedantic.dart';
import 'package:web_socket_channel/io.dart';

class WebSocketChannelWrapper {
  int _requestIds;
  bool _wasOpen = false, _autoReconnect = true;
  Map<dynamic, StreamController> _channels;
  IOWebSocketChannel _socket;
  StreamController _connectStream, _onDoneStream;
  Timer _reconnectTimer, _autoRecTimer;

  final double _RECONNECT_VALUE_MIN = 1000, // 1 second
               _RECONNECT_VALUE_MAX = 1000 * 60.0, // 1 minute
               _RECONNECT_VALUE_FACTOR = 1.4;

  double _reconnectValue;
  var _url, _protocols, _headers, _pingInterval;

  WebSocketChannelWrapper(String url,
      {Iterable<String> protocols, Map<String, dynamic> headers, Duration pingInterval}) : _requestIds = 0,
                                _connectStream = StreamController.broadcast(),
                                _onDoneStream = StreamController.broadcast(),
                                _channels = HashMap()
      {
        _reconnectValue = _RECONNECT_VALUE_MIN;
        _init(url, protocols: protocols, headers: headers, pingInterval: pingInterval);
      }

  set autoReconnect (bool value) => _autoReconnect = value;
  bool get autoReconnect => _autoReconnect;

  _fromSocket(WebSocket socket) => IOWebSocketChannel(socket);

  _init(url, {protocols, headers, pingInterval, connectionTimer}) {
    _url = url;
    _protocols = protocols;
    _headers = headers;
    _pingInterval = pingInterval;

    WebSocket.connect(_url, headers: _headers, protocols: _protocols).then((ws) {
      ws.pingInterval = pingInterval;
      _socket = _fromSocket(ws);
      _socket.stream.listen(_onData, onDone: _onDone);
    }).catchError((e) => _reconnect());
  }

  /// Auto-reconnect using exponential back-off
  _reconnect () {
    if(!_autoReconnect){
      return;
    }

    if(_wasOpen) {
        _reconnectValue = _RECONNECT_VALUE_MIN;
    }else {
        _reconnectValue = min<num>(_reconnectValue * _RECONNECT_VALUE_FACTOR, _RECONNECT_VALUE_MAX);
    }

    _reconnectTimer = Timer(Duration(milliseconds: _reconnectValue.round()), () => 
                          _init(_url, protocols: _protocols, headers: _headers, pingInterval: _pingInterval));
  }

  /// Stream that gets called every time the WebSocket connects
  /// including reconnect
  Stream get onConnect => _connectStream.stream;

  /// Gets called every time the WebSocket disconnect from the server.  
  /// On every call it will try to reconnect
  Stream get onDone => _onDoneStream.stream;

  _socketSend(data, [int reqId]){
    Map<String, dynamic> packet = {'a': data};

    if (reqId != null){
      packet['i'] = reqId;
    }

    _socket.sink.add(json.encode(packet));
  }

  Map _listToMap(String event, List args){
    Map<String, dynamic> data = {'0':event};
    
    args?.asMap()?.forEach((i, e) => data[(++i).toString()] = e);

    return data;
  }

  /// Send [args] to the server with [event] as channel
  emit(String event, [List args]) {
    var data = _listToMap(event, args);
    _socketSend(data);
  }

  /// Send the data extacly like `emit` but could 
  /// recieve response from the server in the form of Future.
  Future request(String event, [List args]) async {
    var data = _listToMap(event, args);
    _socketSend(data, _requestIds);

    Map res = await _on(_requestIds++).first;
    
    var id = res['i'];
    unawaited(_channels[id].close());
    _channels[id] = null;

    var completer = Completer();

    if (res.containsKey('e')){
      completer.completeError(res['e']);
    }else if (res.containsKey('d')){
      completer.complete(res['d']);
    }

    return completer.future;
  }

  Stream _on(key){
    StreamController sc = StreamController();
    _channels[key] = sc;
    return sc.stream;
  }

  /// Returns listener on [event] name 
  Stream on(String event) => _on(event);

  _onData(data) {
    Map map = json.decode(data);
    
    if (map.containsKey('i')){
      _channels[map['i']].sink.add(map);
    }else if (map.containsKey('a')){
      var key = map['a']['0'];

      if (key == 'connect'){
        _wasOpen = true;
        _connectStream.sink.add(null);
        _reconnectTimer?.cancel();
        _autoRecTimer?.cancel();
      }else if (_channels.containsKey(key)){
        _channels[key].sink.add(map['a']['1']);
      }
    }
  }

  _onDone() {
    _reconnect();
    _onDoneStream.sink.add(null);
    _wasOpen = false;
  }

  /// Close the socket and send [closeCode] and [closeReason] to the server
  close({int closeCode, String closeReason}){
    _autoReconnect = false;
    _channels.forEach((_, sc) => sc?.close());
    _socket.sink.close(closeCode ?? 1000, closeReason);
  }
}

example/websocket_channel_wrapper_example.dart
import 'dart:async';
import 'dart:io';
import 'package:websocket_channel_wrapper/websocket_channel_wrapper.dart';

const PORT = 42069;

main() {
  WebSocketChannelWrapper socket = WebSocketChannelWrapper('ws://192.168.0.18:$PORT',
                                              headers: {'id': '1234567890qwertyuiop'});

  print('Connecting...');

  var onReady = socket.onConnect.first;

  onReady.timeout(Duration(seconds: 5), onTimeout: () => print('Server could not be reach, trying to re-connect'));

  socket.onConnect.skip(1).listen((_) => print('Successful re-connection to server'));

  socket.onDone.listen((_) => print(socket.autoReconnect ? 'Connection lost, trying to re-connect!' : 'Connection ended!'));

  onReady.then((_) async { // [onConnect] It's called every time the WebSocket reconnect
    print('Connected!');

    socket.emit('msg', ['DART_WEBSOCKET_WRAPPER', 'Hello, World!']);

    socket.on('serverTime').listen((time) => print('Time: $time'));  // Time: 2019-07-29T14:00:32.635Z

    socket.request('userCount').then((n) => print('# users: $n'));   // # users: 1

    socket.request('checkError').catchError((e) => print(e));        // Yep, errors work

    var filename = '/path/to/file/image.jpg';
    sendFile(filename, socket);   // Send binary file over socket

    Timer(Duration(seconds: 10), () {    // Close the socket after 1 minute
      var reason = 'break time';
      print('Close reason: $reason');
      socket.close(closeReason: reason);
    });
  });
}

sendFile(String filename, socket) async {
  var image = File(filename);
  var contents = await image.readAsBytes();

  final stopwatch = Stopwatch()..start();   // Start timer to measure speed
  
  socket.request('transferFile', [filename, contents]).then((_) { // Send first file name then the bytes

    final speed = contents.length / stopwatch.elapsedMicroseconds;
    print('File sended in ${stopwatch.elapsed} at $speed MB/s');  // File sended in 0:00:00.058191 at 1.63 MB/s

  }).catchError((e) => print(e));
}
